internal/controller/promotion/git.go

```diff
@@ -119,7 +119,7 @@ func (g *gitMechanism) Promote(
 		); err != nil {
 			return nil, newFreight, err
 		}
-		newStatus = aggregateGitPromoStatus(newStatus, *otherStatus)
+		newStatus = aggregateGitPromoStatus(newStatus, *otherStatus).WithFreight(&newFreight)
```

Comment:
I am not sure setting this here is the best thing we can do. As this would mean that for every promotion mechanism which does something with the Freight, we would have to remember to include it in the status.

What I would do instead, is persist the `FreightReference` to the `PromotionStatus` within `internal/controller/promotions/promotions.go` after `r.promoMechanisms.Promote` has been called. As `nextFreight` will contain the updated Freight data.

Reply:
> What I would do instead, is persist the `FreightReference` to the `PromotionStatus` within `internal/controller/promotions/promotions.go` after `r.promoMechanisms.Promote` has been called. As `nextFreight` will contain the updated Freight data.

I see, but is it more convenient to add freight reference right beside the place we update its details? Also, it is pretty unclear which promote mechanism cause freight reference update even if we update `PromotionStatus` in `nternal/controller/promotions/promotions.go`

Reply:
> > I see, but is it more convenient to reflect freight reference changes right beside the place we update its details?
> 
> Can you eleborate more on this question, I am not sure I am following the details of it.
> 


I'm just wondering if it is more clear to add the new detail of FreightReference to the Promotion Status. In that way, after calling r.promoMechanisms.Promote, the caller will have the latest FreightReference data.

> > Also, it is pretty unclear which promotion mechanism caused freight reference update even if we update `PromotionStatus` in `internal/controller/promotions/promotions.go`
> 
> Who and/or why would we have the desire to know what promotion mechanism changed something? Or how is the current approach solving this better?

For the second part, since we're doing an aggregate promote mechanism, will it be useful (or not) to know that which promote mechanism changed the FreightReference detail.

Reply:
I believe the easiest place to assign the FreightReference to the PromotionStatus would be right here:

https://github.com/akuity/kargo/blob/5287d0e47578ff09af88e76eb091f28affdc81b1/internal/controller/promotions/promotions.go#L455-L458

A superior, but more involved option might be to modify the signature of the `Mechanism` interface's `Promote` function so that promotion mechanisms do not return PromotionStatus and modified FreightReference separately, but just return the PromotionStatus with the FreightReference already set.

I think I would suggest taking the easier approach and either @hiddeco or myself will do the rest of the housekeeping in a follow-up.